spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
+
+
+def run_comm_fusion(gm: fx.GraphModule) -> fx.GraphModule:
     """main entry into remapping graph for all_reduce fusion"""
 
-    result = False
+    # result = False  TODO - do we return a success code or the graph,
+    # since graph is modified directly in place?
+
+    # first recompile to make sure we have coherent graph
+    gm.recompile()
 
     # get our main graph info
-    graph_info = GraphInfo()
-    graph_info.update_info(gm)
+    gi = GraphInfo()
+    gi.update_info(gm)
+
+    # _debug(f"\n Start of fusion pass graph {gm.graph.print_tabular()}\n")
 
     # scan graph for all comm sections (fusion elements)
     fe_list = _scan_graph_for_fusion_elements(gm, comm_type=CommType.allreduce)
 
+    gi.num_starting_fe = len(fe_list)  # type: ignore
+    gi.fe_list = fe_list
+
+    # simple fusion policy where int = num buckets to fuse...start with 2,
+    # meaning every 2 comms are fused into 1
+    fusion_policy: int = 2
+
+    # determine peak memory using fusion policy
+    peak_memory_required = _determine_peak_memory(gi, fusion_policy)
+
+    buffer_node = _insert_fusion_buffer_node(gm, peak_memory_required, gi)
+
+    # Main process loop - iterate all fusion elements, apply fusion to subsets
+
+    offset = 0
+    count = 0
+    # new_output_args=[]
+    start_output_args: List[fx.Node] = gi.output.args[0]  # type: ignore
+    new_output_args: List[fx.Node] = list(start_output_args)  # type: ignore
+
+    # ----------- main fusion loop ------------------------
+
+    for index, item in enumerate(gi.fe_list):  # type: ignore
+        count += 1
+        if count == fusion_policy:
+            start_index = offset
+            stop_index = offset + count
+
+            curr_fe_list = gi.fe_list[start_index:stop_index]  # type: ignore
+
+            _copy_fe_to_buffer(gi, gm, curr_fe_list)
+
+            _scatter_results_from_buffer(gi, gm, curr_fe_list)
+
+            # switch wait_comms to output gradient nodes in output directly
+            # fusion will have removed and reworked existing wait_comms
+            # TODO - this will break atm for dynamic fusion...rework for unlimited fusion case.
+            _finalize_output_node(
+                gi, gm, curr_fe_list, start_index, stop_index, new_output_args
+            )
+
+            offset += count
+            count = 0
+    # update output with the updated args
+    gm.graph.erase_node(gi.output)
+    gm.graph.output(new_output_args)
+
+    _debug(f"631, processed {index+1} fe items")
+
+    # final verification of output node - # TODO remove as this is debugging util
+    # for node in reversed(gm.graph.nodes):
+    #    if node.op == OP.OUTPUT:
+    #       new_output = node
+    #       break
+    # _debug(f"703, updated output node args {new_output.args=}\n")
+
     # final review print
     graph_cleanup(gm)
 
-    pretty_print_graph(gm, "final version, fusion pass")
+    # TODO - remove this.  This is purely a final check to verify all meta data
```

Comment:
remove?

Reply:
yes, removed.  I think ultimately we may want to add some generic debug utils in graph_utils as we encounter complex graphs but will do that as functions, instead of the above commented out sections :)